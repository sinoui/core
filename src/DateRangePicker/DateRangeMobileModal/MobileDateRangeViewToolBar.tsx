import Button from '@sinoui/core/Button';
import Caption from '@sinoui/core/Caption';
import H5 from '@sinoui/core/H5';
import IconButton from '@sinoui/core/IconButton';
import Close from '@sinoui/core/svg-icons/Close';
import styled from 'styled-components';

/**
 * 工具栏组件的属性
 */
interface Props {
  /**
   * 弹窗标题
   */
  title?: string;
  /**
   * 开始时间
   */
  startDate?: Date;
  /**
   * 结束时间
   */
  endDate?: Date;
  /**
   * 弹窗关闭时的回调函数
   */
  onClose?: () => void;
  /**
   * 点击保存按钮时的回调函数
   */
  onOk?: () => void;
  /**
   * 点击清除按钮时的回调函数
   */
  onClear?: () => void;
  /**
   * 聚焦的输入框
   */
  focused?: 'start' | 'end';
  /**
   * 焦点位置变更回调函数
   */
  onFocusedChange?: (type: 'start' | 'end') => void;
  /**
   * 开始时间标题
   */
  startTitle?: string;
  /**
   * 结束时间标题
   */
  endTitle?: string;
}

const MobileDateRangeViewToolBarWrapper = styled.div`
  height: 128px;
  padding: 8px 16px 24px 12px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: ${({ theme }) => theme.palette.primary.main};
  color: ${({ theme }) => theme.palette.primary.contrastText};
  box-sizing: border-box;

  & > * {
    color: currentColor;
  }

  .sinoui-icon-button,
  .sinoui-button {
    color: ${({ theme }) => theme.palette.primary.contrastText};
  }

  .sinoui-typography--caption,
  .sinoui-typography--h5 {
    padding-left: 48px;
  }
`;

const MobileDateRangeViewToolBarAction = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

/**
 * 日期包裹组件的属性
 */
interface DateWrapperProps {
  /**
   * 是否是聚焦状态
   */
  $focused?: boolean;
}

const DateWrapper = styled.span<DateWrapperProps>`
  opacity: ${({ $focused }) => ($focused ? 1 : 0.7)};
  cursor: pointer;
`;

/**
 * 日历视图的工具栏
 *
 * @param props 组件属性
 * @param props.title 弹窗标题
 * @param props.startDate 开始时间
 * @param props.endDate 结束时间
 * @param props.onClose 弹窗关闭时的回调函数
 * @param props.onOk 点击保存按钮时的回调函数
 * @param props.onClear 点击清除按钮时的回调函数
 * @param props.focused 聚焦的输入框
 * @param props.onFocusedChange 焦点位置变更回调函数
 * @param props.startTitle 开始时间标题
 * @param props.endTitle 结束时间标题
 */
const MobileDateRangeViewToolBar = ({
  title,
  startDate,
  endDate,
  onClose,
  onOk,
  onClear,
  focused,
  onFocusedChange,
  startTitle = '开始时间',
  endTitle = '结束时间',
  ...rest
}: Props) => {
  const handleFocusedChange = (type: 'start' | 'end') => {
    if (onFocusedChange) {
      onFocusedChange(type);
    }
  };
  return (
    <MobileDateRangeViewToolBarWrapper
      className="sinoui-date-range-view-toolbar"
      {...rest}
    >
      <MobileDateRangeViewToolBarAction>
        <IconButton dense onClick={onClose}>
          <Close />
        </IconButton>
        <MobileDateRangeViewToolBarAction>
          <Button onClick={onClear}>清除</Button>
          <Button onClick={onOk}>保存</Button>
        </MobileDateRangeViewToolBarAction>
      </MobileDateRangeViewToolBarAction>
      <Caption>{title}</Caption>
      <H5 className="sinoui-date-range-view-toolbar__selected-date">
        <DateWrapper
          $focused={focused === 'start'}
          onClick={() => handleFocusedChange('start')}
        >
          {startDate
            ? `${startDate.getMonth() + 1}月${startDate.getDate()}日`
            : startTitle}
        </DateWrapper>
        -
        <DateWrapper
          $focused={focused === 'end'}
          onClick={() => handleFocusedChange('end')}
        >
          {endDate
            ? `
        ${endDate.getMonth() + 1}月${endDate.getDate()}日`
            : endTitle}
        </DateWrapper>
      </H5>
    </MobileDateRangeViewToolBarWrapper>
  );
};

export default MobileDateRangeViewToolBar;
