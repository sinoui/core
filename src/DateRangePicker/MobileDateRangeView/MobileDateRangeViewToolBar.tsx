import React from 'react';
import styled from 'styled-components';
import Caption from '@sinoui/core/Caption';
import H5 from '@sinoui/core/H5';
import IconButton from '@sinoui/core/IconButton';
import Close from '@sinoui/core/svg-icons/Close';
import Button from '@sinoui/core/Button';

interface Props {
  /**
   * 弹窗标题
   */
  title?: string;
  /**
   * 开始时间
   */
  startDate?: Date;
  /**
   * 结束时间
   */
  endDate?: Date;
}

const MobileDateRangeViewToolBarWrapper = styled.div`
  height: 128px;
  padding: 8px 16px 24px 12px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: ${({ theme }) => theme.palette.primary.main};
  color: ${({ theme }) => theme.palette.primary.contrastText};
  box-sizing: border-box;

  & > * {
    color: currentColor;
  }

  .sinoui-icon-button,
  .sinoui-button {
    color: ${({ theme }) => theme.palette.primary.contrastText};
  }

  .sinoui-typography--caption,
  .sinoui-typography--h5 {
    padding-left: 48px;
  }
`;

const MobileDateRangeViewToolBarAction = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

/**
 * 日历视图的工具栏
 */
const MobileDateRangeViewToolBar = ({
  title,
  startDate,
  endDate,
  ...rest
}: Props) => {
  return (
    <MobileDateRangeViewToolBarWrapper
      className="sinoui-date-range-view-toolbar"
      {...rest}
    >
      <MobileDateRangeViewToolBarAction>
        <IconButton dense>
          <Close />
        </IconButton>
        <Button>保存</Button>
      </MobileDateRangeViewToolBarAction>
      <Caption>{title}</Caption>
      <H5>
        {startDate
          ? `${startDate.getMonth() + 1}月${startDate.getDate()}日`
          : '开始时间'}
        --
        {endDate
          ? `
        ${endDate.getMonth() + 1}月${endDate.getDate()}日`
          : '结束时间'}
      </H5>
    </MobileDateRangeViewToolBarWrapper>
  );
};

export default MobileDateRangeViewToolBar;
