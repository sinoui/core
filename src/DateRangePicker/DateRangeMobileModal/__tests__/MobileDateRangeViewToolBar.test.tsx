import React from 'react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { cleanup, render, fireEvent } from '@testing-library/react';
import MobileDateRangeViewToolBar from '../MobileDateRangeViewToolBar';

afterEach(cleanup);

describe('快照测试', () => {
  it('渲染', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <MobileDateRangeViewToolBar />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('验收测试', () => {
  it('显示', () => {
    const { getByText } = render(
      <ThemeProvider theme={defaultTheme}>
        <MobileDateRangeViewToolBar />
      </ThemeProvider>,
    );

    expect(getByText('保存')).toBeInTheDocument();
    expect(getByText('清除')).toBeInTheDocument();
  });

  it('点击清除按钮，清空选中时间', () => {
    const onClear = jest.fn();
    const { getByText } = render(
      <ThemeProvider theme={defaultTheme}>
        <MobileDateRangeViewToolBar
          startDate={new Date(2020, 6, 20)}
          endDate={new Date(2020, 7, 10)}
          onClear={onClear}
        />
      </ThemeProvider>,
    );

    expect(getByText('7月20日')).toBeInTheDocument();

    fireEvent.click(getByText('清除'));
    expect(onClear).toBeCalled();
  });

  it('点击保存按钮，onOk被调用', () => {
    const onOk = jest.fn();
    const { getByText } = render(
      <ThemeProvider theme={defaultTheme}>
        <MobileDateRangeViewToolBar
          startDate={new Date(2020, 6, 20)}
          endDate={new Date(2020, 7, 10)}
          onOk={onOk}
        />
      </ThemeProvider>,
    );
    fireEvent.click(getByText('保存'));
    expect(onOk).toBeCalled();
  });

  it('点击开始时间,onFocusedChange被调用', () => {
    const onFocusedChange = jest.fn();
    const { getByText } = render(
      <ThemeProvider theme={defaultTheme}>
        <MobileDateRangeViewToolBar
          onFocusedChange={onFocusedChange}
          focused="end"
        />
      </ThemeProvider>,
    );
    fireEvent.click(getByText('开始时间'));
    expect(onFocusedChange).toBeCalledWith('start');
    expect(getByText('结束时间')).toHaveStyle('opacity:1');
    expect(getByText('开始时间')).toHaveStyle('opacity:0.7');
  });

  it('onClose属性', () => {
    const onClose = jest.fn();
    const { container } = render(
      <ThemeProvider theme={defaultTheme}>
        <MobileDateRangeViewToolBar onClose={onClose} focused="end" />
      </ThemeProvider>,
    );

    fireEvent.click(
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      container.querySelector(
        '.sinoui-date-range-view-toolbar .sinoui-icon-button',
      )!,
    );

    expect(onClose).toBeCalled();
  });
});
