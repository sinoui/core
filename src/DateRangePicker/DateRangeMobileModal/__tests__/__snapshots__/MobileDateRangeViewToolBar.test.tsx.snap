// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 focused=end 1`] = `
<div
  className="sc-jRQBWg jpEcvy sinoui-date-range-view-toolbar"
>
  <div
    className="sc-gKclnd eCRbOS"
  >
    <button
      className="sc-bdvvtL eqWEUT sinoui-base-button sc-hKwDye iuoEFl sinoui-icon-button sinoui-icon-button--dense"
      color="textSecondary"
      role="button"
      type="button"
    >
      <svg
        aria-hidden="true"
        className="sc-eCImPb gizrW sinoui-svg-icon"
        focusable="false"
        role="presentation"
        size={24}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
    <div
      className="sc-gKclnd eCRbOS"
    >
      <button
        className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ hpPSWF sinoui-button"
        color="primary"
        role="button"
        type="button"
      >
        清除
      </button>
      <button
        className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ hpPSWF sinoui-button"
        color="primary"
        role="button"
        type="button"
      >
        保存
      </button>
    </div>
  </div>
  <p
    className="sc-dkPtRN euzasc sinoui-typography sinoui-typography--caption"
    color="textPrimary"
    type="caption"
  />
  <h5
    className="sc-dkPtRN scNXs sinoui-typography sinoui-typography--h5 sinoui-date-range-view-toolbar__selected-date"
    color="textPrimary"
    type="h5"
  >
    <span
      className="sc-iCfMLu dTLiKD"
      onClick={[Function]}
    >
      开始时间
    </span>
    -
    <span
      className="sc-iCfMLu jgyJvb"
      onClick={[Function]}
    >
      结束时间
    </span>
  </h5>
</div>
`;

exports[`快照测试 focused=start 1`] = `
<div
  className="sc-jRQBWg jpEcvy sinoui-date-range-view-toolbar"
>
  <div
    className="sc-gKclnd eCRbOS"
  >
    <button
      className="sc-bdvvtL eqWEUT sinoui-base-button sc-hKwDye iuoEFl sinoui-icon-button sinoui-icon-button--dense"
      color="textSecondary"
      role="button"
      type="button"
    >
      <svg
        aria-hidden="true"
        className="sc-eCImPb gizrW sinoui-svg-icon"
        focusable="false"
        role="presentation"
        size={24}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
    <div
      className="sc-gKclnd eCRbOS"
    >
      <button
        className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ hpPSWF sinoui-button"
        color="primary"
        role="button"
        type="button"
      >
        清除
      </button>
      <button
        className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ hpPSWF sinoui-button"
        color="primary"
        role="button"
        type="button"
      >
        保存
      </button>
    </div>
  </div>
  <p
    className="sc-dkPtRN euzasc sinoui-typography sinoui-typography--caption"
    color="textPrimary"
    type="caption"
  />
  <h5
    className="sc-dkPtRN scNXs sinoui-typography sinoui-typography--h5 sinoui-date-range-view-toolbar__selected-date"
    color="textPrimary"
    type="h5"
  >
    <span
      className="sc-iCfMLu jgyJvb"
      onClick={[Function]}
    >
      开始时间
    </span>
    -
    <span
      className="sc-iCfMLu dTLiKD"
      onClick={[Function]}
    >
      结束时间
    </span>
  </h5>
</div>
`;

exports[`快照测试 渲染 1`] = `
<div
  className="sc-jRQBWg jpEcvy sinoui-date-range-view-toolbar"
>
  <div
    className="sc-gKclnd eCRbOS"
  >
    <button
      className="sc-bdvvtL eqWEUT sinoui-base-button sc-hKwDye iuoEFl sinoui-icon-button sinoui-icon-button--dense"
      color="textSecondary"
      role="button"
      type="button"
    >
      <svg
        aria-hidden="true"
        className="sc-eCImPb gizrW sinoui-svg-icon"
        focusable="false"
        role="presentation"
        size={24}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
    <div
      className="sc-gKclnd eCRbOS"
    >
      <button
        className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ hpPSWF sinoui-button"
        color="primary"
        role="button"
        type="button"
      >
        清除
      </button>
      <button
        className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ hpPSWF sinoui-button"
        color="primary"
        role="button"
        type="button"
      >
        保存
      </button>
    </div>
  </div>
  <p
    className="sc-dkPtRN euzasc sinoui-typography sinoui-typography--caption"
    color="textPrimary"
    type="caption"
  />
  <h5
    className="sc-dkPtRN scNXs sinoui-typography sinoui-typography--h5 sinoui-date-range-view-toolbar__selected-date"
    color="textPrimary"
    type="h5"
  >
    <span
      className="sc-iCfMLu dTLiKD"
      onClick={[Function]}
    >
      开始时间
    </span>
    -
    <span
      className="sc-iCfMLu dTLiKD"
      onClick={[Function]}
    >
      结束时间
    </span>
  </h5>
</div>
`;
